spec/controllers/shared_upsell_examples.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require "spec_helper"
+
+RSpec.shared_examples_for "the upsell banner is shown" do
+  subject { get :upsell }
```

Comment:
I am pretty unhappy about the naming of this shared example. Reading the specs I was surprised that there is a generic helper to check for the presence of an upsell banner, then I realized that the tests don't even specify which _action_ would show the banner (I was thinking about showing a banner on an edit page or something). Now I see that the action is something that we expect to be present on the tested controllers.

So I guess `it_behaves_like "having an upsell action"` would be my proposal.

I am also wondering whether there is a nicer way to include shared examples (i.e. avoiding `../../../`), but at least it's somewhat clear where the example comes from.

Reply:
Not sure about a nicer way to include these. For now this will have to do. I've addressed the other comment.